beeline.go

```diff
@@ -97,65 +102,136 @@ func Init(config Config) {
+		trace.AddRollupField(namespacedKey, val)
 	}
-	return nil
 }
 
-// contextBuilder isn't used yet but matches ContextEvent. When it's useful,
-// export it, but until then it's just confusing.
-func contextBuilder(ctx context.Context) *libhoney.Builder {
-	if bldr, ok := ctx.Value(honeyBuilderContextKey).(*libhoney.Builder); ok {
-		return bldr
+// AddFieldToTrace adds the field to both the currently active span and all
+// other spans involved in this trace that occur within this process.
+// Additionally, these fields are packaged up and passed along to downstream
+// processes if they are also using a beeline. This function is good for adding
+// context that is better scoped to the request than this specific unit of work,
+// eg user IDs, globally relevant feature flags, errors, etc. Fields added here
+// are prefixed with `global.`
+func AddFieldToTrace(ctx context.Context, key string, val interface{}) {
+	namespacedKey := fmt.Sprintf("global.%s", key)
+	trace := internal.GetTraceFromContext(ctx)
+	if trace != nil {
+		trace.AddField(namespacedKey, val)
 	}
-	return nil
+}
+
+// HasTrace returns true if there is a trace in the current context
+func HasTrace(ctx context.Context) bool {
+	trace := internal.GetTraceFromContext(ctx)
+	return trace != nil
+}
+
+// StartTraceWithIDs lets you start a trace with a specific set of IDs - it is
+// used when you've received the IDs from another source (eg incoming HTTP
+// headers). If you don't care what the IDs are, you may use either this or
+// StartSpan to start a trace. You cannot change the trace IDs after a trace has
+// begun.
+func StartTraceWithIDs(ctx context.Context, traceID, parentID, name string) context.Context {
+	trace := internal.MakeNewTrace(traceID, parentID, name)
+	ctx, _ = internal.PutTraceInContext(ctx, trace)
+	return ctx
+}
+
+// StartSpan lets you start a new span as a child of an already instrumented
+// handler. If there isn't an existing wrapped handler in the context when this
+// is called, it will start a new trace. You may ignore the returned context
+// except when starting a new trace. Spans automatically get a `duration_ms`
+// field when they are ended; you should not explicitly set the duration unless
+// you want to override it. The name argument will be the primary way the span
+// is identified in the trace view within Honeycomb.
+func StartSpan(ctx context.Context, name string) context.Context {
+	return internal.PushSpanOnStack(ctx, name)
+}
+
+// TODO change this to return a span object rather than a libhoney event to make
+// it easier to treat appropriately for sampling
+
+// StartAsyncSpan is different from StartSpan in that it hands back a libhoney
+// event, and it's the caller's responsibility to handle everything from there
+// on - measuring the duration, adding appropriate fields, and sending the
+// event. Unlike normal spans, when finishing a trace, it does not get sent.
+func StartAsyncSpan(ctx context.Context, name string) *libhoney.Event {
```

Comment:
It'd be helpful to be able to create a background context with the event associated with an async span. This way we can just pass around a context instead of a context _and_ an event.

Reply:
Changed this to return a context, same as all the others. This should make it much more consistent to use.